File: 4-manufactured-pointer-dereference/code.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef struct {
    int key;
    char data[32];
    int is_valid;
} CacheEntry;

typedef struct {
    CacheEntry* entries;
    int capacity;
    int* fast_access_table; // Table of offsets for fast lookups
} MemoryCache;

MemoryCache* initializeCache(int capacity) {
    MemoryCache* cache = (MemoryCache*)malloc(sizeof(MemoryCache));
    if (cache == NULL) {
        return NULL;
    }
    
    cache->capacity = capacity;
    cache->entries = (CacheEntry*)malloc(sizeof(CacheEntry) * capacity);
    cache->fast_access_table = (int*)malloc(sizeof(int) * capacity);
    
    if (cache->entries == NULL || cache->fast_access_table == NULL) {
        free(cache->entries);
        free(cache->fast_access_table);
        free(cache);
        return NULL;
    }
    
    for (int i = 0; i < capacity; i++) {
        cache->entries[i].key = -1;
        memset(cache->entries[i].data, 0, 32);
        cache->entries[i].is_valid = 0;
        cache->fast_access_table[i] = i; // Initially just direct mapping
    }
    
    return cache;
}

void addToCacheUnsafe(MemoryCache* cache, int key, const char* data) {
    // Find an empty slot
    int index = key % cache->capacity;
    cache->entries[index].key = key;
    strncpy(cache->entries[index].data, data, 31);
    cache->entries[index].data[31] = '\0';
    cache->entries[index].is_valid = 1;
}

// This function deliberately creates a "manufactured" pointer 
// by converting an arbitrary integer to a pointer
CacheEntry* getDirectPointerUnsafe(int address_value) {
    // Intentionally creating a manufactured pointer
    // 0x1337 is an arbitrary address that is unlikely to be valid
    CacheEntry* manufactured_ptr = (CacheEntry*)address_value;
    return manufactured_ptr;
}

// Attempt to read from cache with direct pointer access (unsafe)
void readWithManufacturedPointer(MemoryCache* cache, int address_value) {
    printf("Attempting to read from manufactured pointer at address 0x%X\n", address_value);
    
    // Get a manufactured pointer to what we "think" is a cache entry
    CacheEntry* entry_ptr = getDirectPointerUnsafe(address_value);
    
    // Dangerous: Dereferencing a manufactured pointer
    printf("Key: %d\n", entry_ptr->key);
    printf("Data: %s\n", entry_ptr->data);
    printf("Is Valid: %d\n", entry_ptr->is_valid);
}

void freeCache(MemoryCache* cache) {
    if (cache != NULL) {
        free(cache->entries);
        free(cache->fast_access_table);
        free(cache);
    }
}

int main() {
    clock_t start, end;
    double cpu_time_used;
    
    start = clock();
    
    MemoryCache* cache = initializeCache(10);
    if (cache == NULL) {
        printf("Failed to initialize cache\n");
        return 1;
    }
    
    addToCacheUnsafe(cache, 123, "Test data 1");
    addToCacheUnsafe(cache, 456, "Test data 2");
    addToCacheUnsafe(cache, 789, "Test data 3");
    
    printf("Cache initialized with 3 entries\n\n");
    
    // First, try a valid access for comparison
    printf("Reading valid cache entry:\n");
    CacheEntry* valid_entry = &(cache->entries[0]);
    printf("Key: %d\n", valid_entry->key);
    printf("Data: %s\n", valid_entry->data);
    printf("Is Valid: %d\n\n", valid_entry->is_valid);
    
    // Now attempt access with manufactured pointers
    // This will likely crash or show garbage data
    printf("Reading with manufactured pointer:\n");
    readWithManufacturedPointer(cache, 0x1337);
    
    // If the program hasn't crashed, try another bad address
    printf("\nReading with another manufactured pointer:\n");
    readWithManufacturedPointer(cache, 0x42424242);
    
    freeCache(cache);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("\nExecution time: %f seconds\n", cpu_time_used);
    
    return 0;
}

File: 4-manufactured-pointer-dereference/code.rs
use std::time::Instant;

struct CacheEntry {
    key: i32,
    data: [u8; 32],
    is_valid: i32,
}

struct MemoryCache {
    entries: Vec<CacheEntry>,
    capacity: usize,
    fast_access_table: Vec<usize>, // Table of offsets for fast lookups
}

impl MemoryCache {
    fn new(capacity: usize) -> Self {
        let mut entries = Vec::with_capacity(capacity);
        let mut fast_access_table = Vec::with_capacity(capacity);
        
        for i in 0..capacity {
            entries.push(CacheEntry {
                key: -1,
                data: [0; 32],
                is_valid: 0,
            });
            fast_access_table.push(i); // Initially just direct mapping
        }
        
        MemoryCache {
            entries,
            capacity,
            fast_access_table,
        }
    }
    
    fn add_to_cache_unsafe(&mut self, key: i32, data: &str) {
        // Find an empty slot
        let index = (key as usize) % self.capacity;
        self.entries[index].key = key;
        
        // Copy data safely (with bounds checking)
        let bytes = data.as_bytes();
        let copy_len = std::cmp::min(bytes.len(), 31);
        self.entries[index].data[..copy_len].copy_from_slice(&bytes[..copy_len]);
        self.entries[index].is_valid = 1;
    }
}

// This function deliberately creates a "manufactured" pointer
// by converting an arbitrary integer to a pointer
fn get_direct_pointer_unsafe(address_value: usize) -> *mut CacheEntry {
    // Intentionally creating a manufactured pointer
    let manufactured_ptr = address_value as *mut CacheEntry;
    manufactured_ptr
}

// Attempt to read from cache with direct pointer access (unsafe)
fn read_with_manufactured_pointer(_cache: &MemoryCache, address_value: usize) {
    println!("Attempting to read from manufactured pointer at address 0x{:X}", address_value);
    
    // Get a manufactured pointer to what we "think" is a cache entry
    let entry_ptr = get_direct_pointer_unsafe(address_value);
    
    // Dangerous: Dereferencing a manufactured pointer
    unsafe {
        println!("Key: {}", (*entry_ptr).key);
        
        // Convert data to a string for printing
        let data_slice = &(*entry_ptr).data;
        let data_str = std::str::from_utf8_unchecked(data_slice);
        println!("Data: {}", data_str);
        
        println!("Is Valid: {}", (*entry_ptr).is_valid);
    }
}

fn main() {
    let start = Instant::now();
    
    let mut cache = MemoryCache::new(10);
    
    cache.add_to_cache_unsafe(123, "Test data 1");
    cache.add_to_cache_unsafe(456, "Test data 2");
    cache.add_to_cache_unsafe(789, "Test data 3");
    
    println!("Cache initialized with 3 entries\n");
    
    println!("Reading valid cache entry:");
    let valid_entry = &cache.entries[0];
    println!("Key: {}", valid_entry.key);
    
    let data_str = std::str::from_utf8(&valid_entry.data).unwrap_or("Invalid UTF-8");
    println!("Data: {}", data_str);
    
    println!("Is Valid: {}\n", valid_entry.is_valid);
    
    // Now attempt access with manufactured pointers
    // This will likely cause a panic in Rust
    println!("Reading with manufactured pointer:");
    read_with_manufactured_pointer(&cache, 0x1337);
    
    // If the program hasn't panicked, try another bad address
    println!("\nReading with another manufactured pointer:");
    read_with_manufactured_pointer(&cache, 0x42424242);
    
    let duration = start.elapsed();
    println!("\nExecution time: {:?}", duration);
}